### PR TITLE
Add filtering to list command logic

### PR DESCRIPTION
**Pull request by Gernene**
This PR adds filtering capabilities to the list command. Please note that some of the methods used in this code may not exist yet.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/ListCommand.java [hunk 1]
```diff
@@ -28,7 +33,7 @@
     /**
      * Creates a ListCommand to list the specified entity, Optionally filtering by other entities
      */
-    public ListCommand(EntityType entityType, Optional<EntityType> filterEntityType,
+    public ListCommand(EntityType entityType, Optional<Entity Type> filterEntityType,
```
Comment (lkldev) [inline src/main/java/seedu/address/logic/commands/ListCommand.java:36]: Seems to be a typo here, should be `EntityType` without space

File: src/main/java/seedu/address/logic/commands/ListCommand.java [hunk 2]
```diff
@@ -41,24 +46,83 @@ public CommandResult execute(Model model) {
         String result = MESSAGE_SUCCESS;
         switch(entityType) {
         case STUDENT:
-            model.updateFilteredStudentList(PREDICATE_SHOW_ALL);
+            updateStudentList(model);
             result = String.format(result, MESSAGE_STUDENTS);
             break;
         case TA_MODULE:
-            model.updateFilteredModuleList(PREDICATE_SHOW_ALL);
```
Comment (lkldev) [inline src/main/java/seedu/address/logic/commands/ListCommand.java:48]: should there be a updateModuleList(model) method called here?

**Review by lkldev — COMMENTED** [review]
some minor comments :) otherwise looks good 👍

File: src/main/java/seedu/address/logic/commands/ListCommand.java [hunk 3]
```diff
@@ -38,27 +45,140 @@ public ListCommand(EntityType entityType, Optional<EntityType> filterEntityType,
...
             break;
         default:
             throw new UnknownEntityException();
         }
         return new CommandResult(result, entityType);
     }
+
+    /**
+     * Updates filtered student list in ModelManager.
+     * @param model ModelManager.
```
Comment (jxt00) [inline src/main/java/seedu/address/logic/commands/ListCommand.java:70]: Do remember to have an empty line between description and parameter section. Same for the other javadocs below.

File: src/main/java/seedu/address/logic/commands/ListCommand.java [hunk 4]
```diff
@@ -38,27 +45,140 @@ public ListCommand(EntityType entityType, Optional<EntityType> filterEntityType,
...
             break;
         case CLASS_GROUP:
-            model.updateFilteredClassGroupList(PREDICATE_SHOW_ALL);
-            result = String.format(result, MESSAGE_CLASS_GROUPS);
+            updateClassGroupList(model, result);
             break;
         case ASSESSMENT:
-            model.updateFilteredAssessmentList(PREDICATE_SHOW_ALL);
-            result = String.format(result, MESSAGE_ASSESSMENTS);
+            updateFilteredAssessmentList(model, result);
```
Comment (jxt00) [inline src/main/java/seedu/address/logic/commands/ListCommand.java:60]: Just curious, why is the method name for `assessment` different from the other entities?

Comment (Gernene) [reply src/main/java/seedu/address/logic/commands/ListCommand.java]: Oops looks like autocomplete. It's fixed now

File: src/main/java/seedu/address/logic/commands/ListCommand.java [hunk 5]
```diff
@@ -38,27 +45,147 @@ public ListCommand(EntityType entityType, Optional<EntityType> filterEntityType,
...
+            } else {
+                result = MESSAGE_INVALID_FILTER;
+            }
+        } catch (IndexOutOfBoundsException e) {
+            result = MESSAGE_INVALID_INDEX;
+        }
+    }
+
+    /**
+     * Updates filtered student list in ModelManager.
```
Comment (EvaderFati) [inline src/main/java/seedu/address/logic/commands/ListCommand.java:96]: Should be class group list here?

**Review by EvaderFati — APPROVED** [review]
Just a minor typo left, others LGTM.

File: src/main/java/seedu/address/logic/commands/ListCommand.java [hunk 6]
```diff
@@ -38,27 +45,147 @@ public ListCommand(EntityType entityType, Optional<EntityType> filterEntityType,
...
         switch(entityType) {
         case STUDENT:
-            model.updateFilteredStudentList(PREDICATE_SHOW_ALL);
-            result = String.format(result, MESSAGE_STUDENTS);
+            updateStudentList(model, result);
             break;
         case TA_MODULE:
-            model.updateFilteredModuleList(PREDICATE_SHOW_ALL);
-            result = String.format(result, MESSAGE_MODULES);
+            updateModuleList(model, result);
```
Comment (lkldev) [inline src/main/java/seedu/address/logic/commands/ListCommand.java:54]: Hmm iirc Strings are immutable in java, hence you wont be able to modify it by pass it into the variable... You would have to return it instead

File: src/main/java/seedu/address/logic/commands/ListCommand.java [hunk 7]
```diff
@@ -38,27 +45,147 @@ public ListCommand(EntityType entityType, Optional<EntityType> filterEntityType,
...
+    }
+
+    /**
+     * Updates filtered assessment list in ModelManager.
+     *
+     * @param model ModelManager.
+     * @param result Result string to update.
+     */
+    private void updateAssessmentList(Model model, String result) {
+        model.updateFilteredAssessmentList(PREDICATE_SHOW_ALL);
```
Comment (lkldev) [inline src/main/java/seedu/address/logic/commands/ListCommand.java:136]: I think we included a filter for assessment as well, as per our user guide

**Review by lkldev — COMMENTED** [review]
Some minor tweaks needed for returning results and filter for assessment

**Review by lkldev — APPROVED** [review]
Thanks! LGTM :)

**Pull request merged**